<h1>Проект по SQL<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта-и-данных" data-toc-modified-id="Описание-проекта-и-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта и данных</a></span><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Схема-днных" data-toc-modified-id="Схема-днных-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Схема днных</a></span></li></ul></li></ul></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Считаем-кол-во-выпущенных-книг-после-1-января-2000-года" data-toc-modified-id="Считаем-кол-во-выпущенных-книг-после-1-января-2000-года-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Считаем кол-во выпущенных книг после 1 января 2000 года</a></span></li><li><span><a href="#Считаем-кол-во-обзоров-и-среднюю-оценку-для-каждой-книги" data-toc-modified-id="Считаем-кол-во-обзоров-и-среднюю-оценку-для-каждой-книги-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Считаем кол-во обзоров и среднюю оценку для каждой книги</a></span></li><li><span><a href="#Находим-издательство,-которое-выпустило-наибольшее-число-книг,-толще-50-страниц" data-toc-modified-id="Находим-издательство,-которое-выпустило-наибольшее-число-книг,-толще-50-страниц-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Находим издательство, которое выпустило наибольшее число книг, толще 50 страниц</a></span></li><li><span><a href="#Находим-автора-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Находим-автора-с-самой-высокой-средней-оценкой-книг-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Находим автора с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Находим-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Находим-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Находим среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Описание проекта и данных

## Описание проекта

Из-за пандемии коронавируса и изменения привычного порядка вещей, люди стали меньше выходить на улицу, чаще оставаясь дома с большим количеством свободного времени, поэтому наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

Наша **задача** как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Чтобы решить поставленную задачу, нам нужно:

- Узнать сколько книг вышло после 1 января 2000 года;
- Определить для каждой книги количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Определить автора с самой высокой средней оценкой книг;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;

- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;

- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;

- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;

- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;

- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Схема днных

![image.png](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

# Подключение к базе данных

In [1]:
# Импортируем необходимые библиотеки

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
from sqlalchemy import create_engine

In [2]:
# Устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

# Изучение данных

In [3]:
# Создадим функцию, которая будет принимать на вход запрос и выводить полученный результат
def query(sql_query):
    return pd.io.sql.read_sql(sql_query, con = engine)

In [4]:
# С помощью цикла изучим все таблицы из базы данных

# Создадим список с названиями таблиц
df_list = ['books','authors','ratings','reviews','publishers']

# Создадим список, который выведет информацию о содержимом для каждой таблицы
for df in df_list:
    
    # Записываем запрос в переменную
    sql = """
              SELECT *
              FROM {}
          """.format(df)
    sql = query(sql)
    
    # Выводим заголовок и содержимое исследуемой таблицы вместе с общей информацией
    print('{0:^125}'.format(f'Обзор данных датасета {df}'))
    display(sql)
    sql.info()
    
    # Отделяем каждый новый датасет друг от друга
    print('-'*100)

                                                 Обзор данных датасета books                                                 


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
----------------------------------------------------------------------------------------------------
                                                Обзор данных датасета authors                                                


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
----------------------------------------------------------------------------------------------------
                                                Обзор данных датасета ratings                                                


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
----------------------------------------------------------------------------------------------------
                                                Обзор данных датасета reviews                                                


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
----------------------------------------------------------------------------------------------------
                                              Обзор данных датасета publishers                                               


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
----------------------------------------------------------------------------------------------------


**Вывод:**

- Мы провели первичный обзор каждой таблицы из базы данных, также посмотрели на общую информацию по каждой из них;
- Все данные полные, без пропусков в столбцах.

# Исследовательский анализ данных

## Считаем кол-во выпущенных книг после 1 января 2000 года

In [5]:
sql_query_1 = """
                  SELECT COUNT(*) AS books_number
                  FROM books
                  WHERE publication_date::date > '2000-01-01'
                  
              """

query(sql_query_1)

books_number
0           819

**Вывод:** 

После 1 января 2000 года вышло `819` книг.

## Считаем кол-во обзоров и среднюю оценку для каждой книги 

In [6]:
sql_query_2 = """
                  SELECT b.title,
                         COUNT(DISTINCT rv.review_id) AS reviews_number,
                         ROUND(AVG(rt.rating),2) AS avg_rating
                  FROM books AS b
                  LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
                  LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
                  GROUP BY 1
                  ORDER BY 2 DESC
                  
              """

query(sql_query_2)

title  reviews_number  \
0                                  Memoirs of a Geisha               8   
1                              Twilight (Twilight  #1)               7   
2    Harry Potter and the Chamber of Secrets (Harry...               6   
3                                     The Glass Castle               6   
4                                      Eat  Pray  Love               6   
..                                                 ...             ...   
994    Anne Rice's The Vampire Lestat: A Graphic Novel               0   
995                            The Natural Way to Draw               0   
996   The Cat in the Hat and Other Dr. Seuss Favorites               0   
997                          Essential Tales and Poems               0   
998  Disney's Beauty and the Beast (A Little Golden...               0   

     avg_rating  
0          4.14  
1          3.66  
2          4.29  
3          4.21  
4          3.40  
..          ...  
994        3.67  
995        3.00  
996        5.00  
997        4.00  
998        4.00  

[999 rows x 3 columns]

**Вывод:**

Больше всего обзоров у книги `Memoirs of a Geisha` - `8`, далее идет `Twilight` - `7` и замыкает первую тройку `Harry Potter and the Chamber of Secrets` с `6` обзорами.

## Находим издательство, которое выпустило наибольшее число книг, толще 50 страниц

In [7]:
sql_query_3 = """
                  SELECT p.publisher,
                         COUNT(b.book_id) AS books_number
                  FROM books AS b
                  LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
                  WHERE b.num_pages > 50
                  GROUP BY 1
                  ORDER BY 2 DESC
                  LIMIT 1
                  
              """

query(sql_query_3)

publisher  books_number
0  Penguin Books            42

**Вывод:**

Издательство, которое выпустило наибольшее число книг толще 50 страниц - `Penguin Books`(42 книги).

## Находим автора с самой высокой средней оценкой книг

In [8]:
sql_query_4 = """ 
                    SELECT author,
                           ROUND(AVG(avg_rating), 2) AS avg_rating
                    FROM
                      (SELECT a.author,
                              b.book_id,
                              ROUND(AVG(rt.rating), 2) AS avg_rating
                       FROM books AS b
                       LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
                       LEFT JOIN authors AS a ON b.author_id = a.author_id
                       GROUP BY 1,
                                2
                       HAVING COUNT(rt.rating) >= 50
                       ) AS books_with_50_ratings
                    GROUP BY 1
                    ORDER BY 2 DESC
                    LIMIT 1    
             """

query(sql_query_4)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**Вывод:**

`J.K. Rowling/Mary GrandPré` - автор с самой высокой средней оценкой книг (с 50 и более оценками) и в среднем составляет - `4.29`.

## Находим среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [9]:
sql_query_5 = """
                 WITH num_of_reviews_per_user AS
                     (SELECT username,
                            COUNT(review_id) AS reviews_number
                     FROM reviews
                     WHERE username IN (SELECT username
                                        FROM ratings
                                        GROUP BY 1
                                        HAVING COUNT(rating) > 50)
                     GROUP BY 1)
                  
                 SELECT ROUND(AVG(reviews_number))::integer AS avg_reviews
                 FROM num_of_reviews_per_user
              """

query(sql_query_5)

avg_reviews
0           24

**Вывод:**

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет - `24`

# Вывод

В ходе проведенного исследования и поставленной перед нами задачей от руководства мы:

1. Подключились к базе данных;
2. Первично ознакомились и изучили структуру таблиц с необходимой для нас информацией;
3. Ответили на поставленные вопросы и сделали соответствующие выводы:
    
    - После 1 января 2000 года вышло `819` книг;
    
    - Больше всего обзоров у книги `Memoirs of a Geisha` - `8`, далее идет `Twilight` - `7` и замыкает первую тройку `Harry Potter and the Chamber of Secrets` с `6` обзорами;
    - Издательство, которое выпустило наибольшее число книг толще 50 страниц - `Penguin Books`(42 книги).
    - `J.K. Rowling/Mary GrandPré` - автор с самой высокой средней оценкой книг (с 50 и более оценками) и в среднем составляет - `4.29`.
    - Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет - `24`